### PR TITLE
fix!: replace dag walkers with generic CID extraction from blocks

### PR DESCRIPTION
**Pull request by achingbrain**
Replace the codec-specific `.dagWalkers` property with a generic dag walker internally that uses the `Block` interface from the `multicodecs` module.

- Removes the `.dagWalkers` property from the Helia interface
- Adds `getCodec` and `getHasher` to retrieve codecs and hashers by code
- Adds `loadCodec` and `loadHasher` options to allow sync or async loading of extra codecs/hashes in addition to staticlly configured ones in the `codecs`/`hashers` keys

BREAKING CHANGE: the `.dagWalkers` property has been removed

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if necessary (this includes comments as well)
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: packages/utils/test/fixtures/create-dag.ts [hunk 1]
```diff
@@ -58,33 +58,67 @@ export interface DAGNode {
...
 
-    dag[name].links.push(subChild)
+    for (const child of parent.links) {
+      if (child.links.length > 0) {
+        await write(child)
+      }
 
-    await addChildren(subChild, name, level + 1, index + i, depth - 1, children, dag, codec, blocks)
+      links.push(
+        await createAndPutBlock(dagCbor.code, dagCbor.encode(child), blocks)
```
Comment (SgtPooki) [inline packages/utils/test/fixtures/create-dag.ts:105]: I know this is just a test fixture, but do we want to remove the ability to use any codec?

File: packages/utils/test/utils/networked-storage.spec.ts [hunk 2]
```diff
@@ -39,9 +39,8 @@ describe('networked-storage', () => {
       logger: defaultLogger(),
       blockBrokers: [
         bitswap
-      ],
-      hashers: defaultHashers()
-    })
+      ]
+    }, getHasher())
```
Comment (SgtPooki) [inline packages/utils/test/utils/networked-storage.spec.ts:43]:
```suggestion
      ],
      getHasher
    })
```

**Review by SgtPooki — CHANGES_REQUESTED** [review]
a few things that could be changed, not sure if the networked-storage accepting an additional arg was done explicitly for some reason though

File: packages/car/src/index.ts [hunk 3]
```diff
@@ -221,18 +218,15 @@ class DefaultCar implements Car {
...
-      throw new Error(`No dag walker found for cid codec ${cid.code}`)
-    }
+    const codec = await this.components.getCodec(cid.code)
+    const bytes = await this.components.blockstore.get(cid, options)
 
-    const block = await this.components.blockstore.get(cid, options)
+    await withBlock(cid, bytes)
 
-    await withBlock(cid, block)
+    const block = createUnsafe({ bytes, cid, codec })
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/car/src/index.ts:226]: We need a `Block` instance to call `.links()` on it.

`createUnsafe` doesn't verify that the block hash matches the block data, which is a (potentially) expensive operation, it just creates the block.

File: packages/utils/test/fixtures/create-dag.ts (hunk 1 above)
Comment (achingbrain) [reply packages/utils/test/fixtures/create-dag.ts:105]: The problem is creating a layered DAG depends on the codec - different codecs will implement this in different ways.

Previously we were only ever using one codec so it was a bit misleading.

File: packages/car/src/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (SgtPooki) [reply packages/car/src/index.ts:226]: For the sake of verified-fetch, where do we verify the block hash matches the data after the changes in this PR?

**Comment by SgtPooki** [comment]
we need to run benchmarks to ensure no perf regression

File: packages/car/src/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/car/src/index.ts:226]: The blockstore should only give us the bytes that match the hash so there's no change necessary there.

**Comment by achingbrain** [comment]
Perf seems fine - the benchmark added by https://github.com/ipfs/helia/pull/633 shows comparable performance between this branch and main.

**Pull request merged**